### ISSUE TITLE
[Bug] MOSS-Transcribe-Diarize 在配置 131K 上下文和 65K 最大输出后，40 分钟音频仍被提前截断

### ISSUE THREAD

**Issue by GLucky-98** [OpenMOSS/MOSS-Transcribe-Diarize#26]
感谢开源！

# [Bug] MOSS-Transcribe-Diarize 在配置 131K 上下文和 65K 最大输出后，40 分钟音频仍被提前截断

## 问题概述

使用 SGLang Omni 部署 `OpenMOSS-Team/MOSS-Transcribe-Diarize` 时，转写 40 分钟音频会在输入结束前数分钟停止。

服务返回 HTTP 200，响应中的输入时长也是完整的 2400 秒，但生成的转写内容在中途结束。使用两段互不相关且语音密集的音频都可以复现，因此问题不像是由某个特定音频的静音或文件损坏造成的。

从现象看，调用链路中可能仍有接近 16K 的输出 token 上限，但当前配置的 `max_new_tokens` 已经是 65536。这里的 16K 只是根据多次结果作出的推测，尚未通过服务返回的 token 用量直接确认。

## 运行环境

- 模型：`OpenMOSS-Team/MOSS-Transcribe-Diarize`
- SGLang：`0.5.12.post1`
- SGLang Omni commit：`f916b86cb324f21d479e798ad36e2a992cfd010a`
- Python：`3.12`
- GPU：单张 NVIDIA H20，显存约 96 GiB
- 启动日志显示的 KV Cache 容量：`691777` tokens

## 启动命令

```bash
HF_HUB_OFFLINE=1 TRANSFORMERS_OFFLINE=1 \
/path/to/sgl-omni serve \
  --model-path /path/to/MOSS-Transcribe-Diarize \
  --port 8000 \
  --max-running-requests 8 \
  --cuda-graph-max-bs 8 \
  --mem-fraction-static 0.80 \
  --stages.asr.factory_args.context_length=131072 \
  --stages.asr.factory_args.max_new_tokens=65536 \
  --log-level debug
```

启动时打印的合并配置可以确认这两个参数已经被解析：

```yaml
factory_args:
  context_length: 131072
  max_new_tokens: 65536
server_args_overrides:
  max_running_requests: 8
  cuda_graph_max_bs: 8
```

## 请求方式

```bash
curl http://127.0.0.1:8000/v1/audio/transcriptions \
  -F model="OpenMOSS-Team/MOSS-Transcribe-Diarize" \
  -F file="@40min_16k_mono.wav" \
  -F response_format="verbose_json" \
  -F temperature="0" \
  -F max_new_tokens="65536"
```

ASR stage 已经配置 `max_new_tokens=65536`，请求中也显式传入了相同的值，但转写仍然被截断。

## 复现结果

两段输入音频的时长都是 2400 秒，格式均为 16 kHz、单声道，并且在返回结果的截断点之后仍然包含语音。

| 输入 | HTTP 状态 | 请求耗时 | 响应分段数 | 文本字符数 | 最后完整时间戳 | 未转写音频长度 |
|---|---:|---:|---:|---:|---:|---:|
| 录音 A | 200 | 103.25 秒 | 526 | 39,526 | 2079.36 秒 | 320.64 秒 |
| 录音 B | 200 | 94.92 秒 | 447 | 44,489 | 2133.63 秒 | 266.37 秒 |

录音 B 的最后一个完整分段结束于 `2133.63s`，原始响应的 `text` 字段停在一个未完成的时间戳标记中：

```text
...[213
```

返回的 JSON 本身可以正常解析，但其中的转写内容不完整。

## 音频尾部验证

对录音 B 从 `2133.63s` 到 `2400s` 的未转写部分进行了独立检查：

- 平均音量：`-26.5 dB`
- 最大音量：`-8.2 dB`
- 参考转写在截断点之后覆盖约 `251.14s` 的语音
- 截断点之后的参考文本包含 `4,842` 个字符和 `17` 个源分段

因此，模型并不是因为剩余音频静音或缺失而停止生成。

## 预期行为

当 `context_length=131072` 且 `max_new_tokens=65536` 时，服务应当完成整段 2400 秒音频的转写。如果命中了其他长度限制，API 至少应该返回明确的长度限制结束原因或错误，而不是把不完整的结果作为正常响应返回。

## 实际行为

- 请求返回 HTTP 200。
- 响应报告的输入时长为 `duration: 2400.0`。
- 根据音频内容密度不同，转写在约 2080 至 2134 秒处停止。
- 原始文本可能在时间戳标记或句子中间结束。
- 转写响应没有提供生成 token 数或 `finish_reason`。

## 希望确认的问题

1. 音频转写路由、请求参数模型、pipeline stage 或 SGLang sampling parameters 中是否存在默认的 `16384` 生成上限？
2. `/v1/audio/transcriptions` 是否会把 `max_new_tokens` 传递到 MOSS 转写 stage 及最终的 SGLang sampling parameters？
3. 是否还需要其他参数，才能将实际输出上限提高到 16K 以上？
4. 转写响应是否可以增加 `finish_reason` 和生成 token 用量，以便区分长度截断与模型正常输出 EOS？

如有需要，我可以进一步提供 debug 日志或可公开复现的音频样本。

**Comment by zimoCCC** [comment]
您好，感谢反馈，您所提及的问题我们已经收到过同样的反馈，可否提供一下截断音频以供复现？我们正在修复这个问题

**Comment by GLucky-98** [comment]
> 您好，感谢反馈，您所提及的问题我们已经收到过同样的反馈，可否提供一下截断音频以供复现？我们正在修复这个问题

能否给我一个邮箱，我可以发送一个稳定复现的例子

**Comment by hasadata** [comment]
This is also reproducible on vLLM. Passing a 1h 04min audio file with max_tokens set to 131 always returns only 33 minutes of transcription, regardless of the configuration.

**Comment by Turn10Innovations** [comment]
<html>
<body>
<p>Adding measurements that I think pin this down more precisely than &quot;somewhere around 16K&quot;, and
suggest it is <strong>learned behaviour in the weights rather than a cap in the call chain</strong>.</p>
<p><strong>Setup:</strong> single RTX A4000 (16 GB), <code>transformers</code> + <code>AutoModelForCausalLM</code>, <code>trust_remote_code=True</code>,
bf16, <code>do_sample=False</code>, <code>max_new_tokens=32768</code>. English two-party and multi-party meeting audio,
16 kHz mono. No vLLM, no SGLang, so this is a third serving path reproducing it.</p>

input audio | output tokens | transcript stops at | audio covered
-- | -- | -- | --
30 min | 12,518 | 30.0 min | 100% (ran out of audio, not tokens)
45 min | 16,336 | 39.1 min | 87%
60 min | 16,339 | 39.1 min | 87%
82 min | 16,337 | 41.3 min | 50%


<p><strong>Three different input lengths, three different audio spans, halting within 3 tokens of each other.</strong></p>
<p>Why I think this is not a configuration or runtime cap:</p>
<ol>
<li><strong><code>16384</code> appears nowhere in the config.</strong> Not in <code>config.json</code>, not in the modeling source.
<code>generation_config.json</code> sets <code>max_new_tokens: 5120</code>, which is <em>lower</em> than the observed stop
and is clearly not what is binding either.</li>
<li><strong>A hard cap would truncate at one fixed integer.</strong> These stop at 16,336 / 16,339 / 16,337,
three <em>different</em> values, all a few dozen tokens below 2^14. A <code>if len(out) &gt;= N: break</code>
anywhere in the call chain produces the same number every time. Variance of this shape looks
like a model that has learned &quot;sequences end around here&quot;, landing wherever the next EOS is
most probable.</li>
<li><strong>Context is not the constraint.</strong> On the 82-minute file the 131k context measured only
<strong>49.8% utilised</strong>, roughly 795 input tokens per audio-minute for this speech density. There
is ample room on both sides.</li>
<li><strong>It stops mid-sentence</strong>, with 21 minutes of audio remaining, and emits a well-formed final
segment rather than any error. Nothing in the return value indicates truncation occurred.</li>
</ol>
<p>Point 4 is the part I would flag hardest for other users: <strong>the failure is completely silent.</strong>
The 82-minute run returned exit code 0, 379 plausible segments and correct speaker labels while
missing half the meeting. There is no signal to branch on. Anyone building on this needs their own
assertion that the last timestamp reaches the end of the input, we added one after being caught by
exactly this.</p>
<p><strong>Question:</strong> was the training data capped at ~16k output tokens per example?
If so that would explain both the value and the variance, and it would mean the fix is retrained
weights rather than anything configurable, which changes what downstream users should build in the
meantime. A clear &quot;yes, chunk for now&quot; would be genuinely useful even without a timeline.</p>
<p>Thanks for the model! The joint
transcribe+diarize output is a large improvement over running ASR and diarization separately, which
is why the long-audio path matters.</p>

</body>
</html>

**Comment by Athena-I** [comment]
模型文件generation_config.json配置中设置了 `max_tokens: 5120`，这个值覆盖了 vLLM 的默认设置，重新启动 vLLM 时添加 `--generation-config vllm` 参数，这样 vLLM 会使用自己的默认参数而不是模型的配置

**Comment by GLucky-98** [comment]
> 模型文件generation_config.json配置中设置了 `max_tokens: 5120`，这个值覆盖了 vLLM 的默认设置，重新启动 vLLM 时添加 `--generation-config vllm` 参数，这样 vLLM 会使用自己的默认参数而不是模型的配置

谢谢你的回复，但是我遇到的问题是截断在16384，而且是正常停止不是因为触及上限停止，而且在sglang和vllm部署的模型上都遇到了同样的问题

**Comment by Lucklady** [comment]
为何判定为模型缺陷
我们已按官方设置 max_new_tokens=65536，仍在 ~16384 停。
diag_moss_eos_probe：step=16383 时 <|im_end|> 概率约 0.59（argmax）——模型自己选 EOS。
禁掉 EOS 后能继续生成合法时间戳到 ~90% 覆盖；说明不是框架硬截断，而是学到的早停行为。
上游推理路径同样是标准 model.generate() + eos_token_id，没有额外防护——HF 路径遇到同类长 ICSI 会同样早停。

**Comment by rodchang7** [comment]
我也遇到了一樣的問題，但我試了最新的vllm版本，這問題似乎解決了，以下是我的設定，供參考。
P.S. 我使用H100 GPU顯卡
Dockerfile:
FROM vllm/vllm-openai:v0.27.1
WORKDIR /vllm-workspace
RUN uv pip install --system -U vllm --torch-backend=auto --extra-index-url https://wheels.vllm.ai/68b4a1d582818e67adc903bf1b8fc5a5447da2fa/cu132
RUN uv pip install --system -U vllm[audio]
RUN apt-get update && apt-get install -y git
RUN git clone https://github.com/OpenMOSS/MOSS-Transcribe-Diarize.git && cd MOSS-Transcribe-Diarize && pip install -e .
RUN uv pip uninstall --system -y torchaudio
RUN uv pip install --system -U --pre torchaudio --index-url https://download.pytorch.org/whl/nightly/cu132

docker-compose.yaml:
services:

  vllm_chat:
    image: vllm/vllm-openai:MOSS_Trnascribe_v0.27.1
    environment:
      - PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
      - VLLM_MAX_AUDIO_DECODE_DURATION_S=5400
      - VLLM_MAX_AUDIO_CLIP_FILESIZE_MB=200
    ports:
      - 8000:8000
    #profiles: [vllm-chat]
    volumes:
      - /home/user123/MOSS-Transcribe-Diarize:/root/MOSS-Transcribe-Diarize
    entrypoint: vllm serve --model /root/MOSS-Transcribe-Diarize --served-model-name MOSS-Transcribe-Diarize --trust-remote-code --dtype bfloat16 --max_num_seqs 4 --cudagraph_capture_sizes 4 --host 0.0.0.0 --port 8000 --max_model_len 131072 --gpu_memory_utilization 0.3 --max-num-batched-tokens 131072
    restart: always
    deploy:
          resources:
            reservations:
              devices:
                - driver: nvidia
                  device_ids: ['0']
                  capabilities: [gpu]

**Comment by helixzz** [comment]
补充一个来自 vLLM 路径的、已做单变量对照的同形案例。为保护会议内容，以下只提供结构和数值，不包含音频、转写正文或业务标识。

### 环境

- 模型：`OpenMOSS-Team/MOSS-Transcribe-Diarize`
- vLLM：`0.23.1rc1.dev949+g68b4a1d58`
- `trust_remote_code=true`，上下文 131,072
- 输入：47:21.192（2,841.192 秒）MP3，16 kHz，单声道，8,523,800 bytes
- 请求：`response_format=json`、`temperature=0`

### 现象

- HTTP 200，`usage={"type":"duration","seconds":2842}`，说明服务端解码到了完整输入时长。
- 原始 `text` 为 21,499 字符；最后一个完整时间戳为 1,811.78 秒（约 30:11）。
- 原始文本没有以闭合的数值结束时间戳收尾；不是下游 parser 丢段。产品 parser 得到 502 段，最大 `end=1,811.28s`，与原始响应上界一致。
- 独立的带时间戳参考转写显示 30:11 后持续有语音，直到约 47:12。对 29:30-31:00 音频做 `silencedetect`，没有持续 1 秒以上的静音。

### 预算对照

同一音频、同一模型、同一温度，分别请求：

| 字段 | 原始响应结果 |
|---|---|
| `max_completion_tokens=48660` | 1,811.78s；response SHA-256 相同 |
| `max_completion_tokens=90000` | 1,811.78s；response SHA-256 相同 |
| `max_completion_tokens=91435` + `max_new_tokens=131072` | 1,811.78s；response SHA-256 相同 |

`91,435 = 131,072 - 37,589（输入估算）- 2,048（预留开销）`，已经是这个输入下的计算生成上界。三份响应逐字节一致，因此本案例不是请求侧输出预算耗尽。

另外，所部署 vLLM wheel 的 `TranscriptionRequest` 只声明 `max_completion_tokens`，没有声明 `max_new_tokens`。仅发送 `max_new_tokens=90000` 时会回落到另一条默认路径，只覆盖到 605.81 秒；不能把它解释为提高了有效上限。

### 判断

这个案例与本 issue 不只是“结果都比较短”：完整输入时长、HTTP 200、截断后仍有持续语音、原始输出没有闭合当前分段、提高到上下文可用上界仍逐字节不变，这些特征都一致。并且它发生在 vLLM，而 issue 首报发生在 SGLang；结合评论中 Transformers 路径约 16,33x token 主动选择 EOS 的探针结果，我们认为这是同一个模型级 learned early-EOS 故障类，而不是某一个 serving backend 的硬截断。由于本次 vLLM 响应不含 token IDs / EOS logits，我们不能仅凭现有数据声称内部 token 位置也已独立复核为完全相同。

当前 vLLM transcription 响应没有 generated-token count、token IDs 或 `finish_reason`，所以我们无法对这一次请求直接复核“EOS 恰好在约 16,384 token”这一内部机制。请问目前是否已有修正版权重、推荐的 EOS 约束方案，或能安全区分“真实音频结束”和“长音频误 EOS”的推理参数？简单禁用 EOS / 强制 `min_new_tokens` 可能在真实尾部静音时生成伪内容，因此我们暂未作为生产修复采用。

**Comment by GLucky-98** [comment]
> 为何判定为模型缺陷 我们已按官方设置 max_new_tokens=65536，仍在 ~16384 停。 diag_moss_eos_probe：step=16383 时 <|im_end|> 概率约 0.59（argmax）——模型自己选 EOS。 禁掉 EOS 后能继续生成合法时间戳到 ~90% 覆盖；说明不是框架硬截断，而是学到的早停行为。 上游推理路径同样是标准 model.generate() + eos_token_id，没有额外防护——HF 路径遇到同类长 ICSI 会同样早停。

多次测试后发现在16384 token的位置 提前截断的概率非常高

**Comment by Hana19951208** [comment]
大佬，现在有解决吗，我现在在sglang中也遇到了这个问题，这是模型训练时的问题吗？这个要咋解决啊

**Comment by Petriorlz** [comment]
> 大佬，现在有解决吗，我现在在sglang中也遇到了这个问题，这是模型训练时的问题吗？这个要咋解决啊

我们会在下一版解决这个问题
